source/lib/vagrant-openstack-provider/client/nova.rb

```diff
@@ -175,6 +196,63 @@ def attach_volume(env, server_id, volume_id, device = nil)
         end
       end
 
+      # List snapshot images associated with a particular server
+      #
+      # @param env [Hash] Vagrant action environment
+      # @param server_id [String] Server UUID
+      #
+      # @return [Array<VagrantPlugins::Openstack::Domain::Image>]
+      def list_snapshots(env, server_id)
+        get_all_images(env, params: { server: server_id })
+      end
+
+      # Create a named snapsot for a given VM
+      #
+      # @param env [Hash] Vagrant action environment
+      # @param server_id [String] Server UUID
+      # @param snapshot_name [String]
+      #
+      # @return [void]
+      def create_snapshot(env, server_id, snapshot_name)
+        instance_exists do
+          post(
+            env,
+            "#{@session.endpoints[:compute]}/servers/#{server_id}/action",
+            { createImage: {
+              name: snapshot_name,
+              metadata: { vagrant_snapshot: true }
```

Comment:
I've been running a patched provider that includes these changes against Kilo for a few months now. I don't have another OpenStack installation available to test against, but from the error message it seems like the API schema for metadata may have changed to allow only strings --- I'll see if I can track that down.